nextcord/voicerec/core.py

```diff
@@ -0,0 +1,338 @@
+"""
+The MIT License (MIT)
+
+:copyright: (c) 2021-present Nextcord Developers
+
+Permission is hereby granted, free of charge, to any person obtaining a
+copy of this software and associated documentation files (the "Software"),
+to deal in the Software without restriction, including without limitation
+the rights to use, copy, modify, merge, publish, distribute, sublicense,
+and/or sell copies of the Software, and to permit persons to whom the
+Software is furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+DEALINGS IN THE SOFTWARE.
+"""
+
+
+import logging
+from io import BufferedWriter, BytesIO
+from select import select
+from struct import pack, unpack_from
+from threading import Thread
+from time import perf_counter, sleep, time as clock_timestamp
+from typing import Optional, Union
+
+from nextcord import File, StageChannel, VoiceChannel, VoiceClient
+from nextcord.client import Client
+
+from . import decrypter, exporters, opus
+from .errors import *
+from .shared import *
+
+ConnectableVoiceChannels = Union[VoiceChannel, StageChannel]
+
+
+AUDIO_HZ = opus.DecoderThread.SAMPLING_RATE
+SILENCE_FRAME_SIZE = 960
+FRAME_OF_SILENCE = b"\xf8\xff\xfe"
+DIFFERENCE_THRESHOLD = 60
+
+STUB = ()
+
+
+SILENCE_STRUCT = pack("<h", 0) * SILENCE_FRAME_SIZE
+SILENCE_STRUCT_10 = SILENCE_STRUCT * 10
+SILENCE_STRUCT_100 = SILENCE_STRUCT * 100
+SILENCE_STRUCT_1000 = SILENCE_STRUCT * 1000  # 3.4mb in size
+
+
+class Silence:
+    def __init__(self, frames: int) -> None:
+        self.frames: int = frames
+
+    def write_to(self, buffer: Union[BufferedWriter, BytesIO]) -> None:

```

Comment:
buffer should be typed as `io.BufferedIOBase`

Reply:
Thank you, will get to that when I work on it again

Reply:
closing, reversed change due to incorrect typing (missing methods used in io.BufferedIOBase